packages/snap/src/index.tsx

```diff
@@ -0,0 +1,569 @@
+          </Section>
+        </Box>
+      ),
+    },
+  });
+}
+
+/**
+ * Sign a deploy.
+ *
+ * @param deployJson - JSON formatted deploy.
+ * @param transaction - JSON formatted transaction.
+ * @param origin - Origin of the request.
+ * @param addressIndex - Address index.
+ * @returns A signed deploy or an error.
+ */
+async function sign(
+  deployJson: any,
+  transaction: any,
+  origin: string,
+  addressIndex = 0,
+) {
+  const publicKeyHex = (await getCSPRAddress(addressIndex)).publicKey;
+  if (!publicKeyHex) {
+    return { error: `Unable to get public key at index ${addressIndex}.` };
+  }
+  try {
+    let transactionParsed;
+    if (deployJson && transaction) {
+      return {
+        error:
+          'Only deployJson field or transaction field should be populated not both.',
+      };
+    }
+    if (deployJson) {
+      if (Transaction.fromJSON(deployJson).getDeploy() === undefined) {
+        return { error: `The field deployJson only supports legacy deploys` };
+      }
+      transactionParsed = Transaction.fromJSON(deployJson);
+    } else {
+      transactionParsed = Transaction.fromJSON(transaction);
+    }
+
+    const bip44Node = await snap.request({
+      method: 'snap_getBip44Entropy',
+      params: {
+        coinType: 506,
+      },
+    });
+    const bip44Nodeaddr = await getBIP44AddressKeyDeriver(bip44Node);
+    const addressKey = await bip44Nodeaddr(addressIndex);
+    const response = await promptUserDeployInfo(
+      transactionParsed,
+      publicKeyHex,
+      origin,
+    );
+    if (!response) {
+      return false;
+    }
+
+    if (addressKey.privateKeyBytes) {
+      if (addressKey.curve === 'ed25519') {
+        const privateKey = PrivateKey.fromHex(
+          addressKey.privateKey?.slice(2) ?? '',
+          KeyAlgorithm.ED25519,
+        );
+        transactionParsed.sign(privateKey);
+        if (deployJson) {
+          return {
+            deploy: Deploy.toJSON(transactionParsed.getDeploy() as Deploy),
+          };
+        }
+        return {
+          transaction: TransactionV1.toJSON(
+            transactionParsed.getTransactionV1() as TransactionV1,
+          ),
+        };
+      }
+
+      if (addressKey.curve === 'secp256k1') {
+        const privateKey = PrivateKey.fromHex(
+          addressKey.privateKey?.slice(2) ?? '',
+          KeyAlgorithm.SECP256K1,
+        );
+        transactionParsed.sign(privateKey);
+        if (deployJson) {
+          return {
+            deploy: Deploy.toJSON(transactionParsed.getDeploy() as Deploy),
+          };
+        }
+        return {
+          transaction: TransactionV1.toJSON(
+            transactionParsed.getTransactionV1() as TransactionV1,
+          ),
+        };
+      }
+
+      return {
+        error: `Unsupported curve : ${addressKey.curve}. Only Secp256K1 && Ed25519 are supported.`,
+      };
+    }
+  } catch (error) {
+    return {
+      error: `Unable to convert json into deploy object.`,
```

Comment:
@KillianH this error message is not accurate in many cases. Can you add `error.message` to the returned error value?

Reply:
Will do that !